### ISSUE TITLE
Scriptatone™ - Fast development of complex behaviors for Baritone

### ISSUE THREAD

**Issue by SIGSTACKFAULT** [cabaletta/baritone#881]
Hook up a scripting language to Baritone, to allow "normal people" to make baritone do complicated things -- no compiling in new commands.

I think ([Jython](https://www.jython.org)) is the best choice, but ([luaj](https://github.com/luaj/luaj)) is also available.

No idea if this is an even vaguely good idea. Maybe it should be a wrapper or something.

**Comment by 0-x-2-2** [comment]
Scripting is a even worse then what we have already because the game is obfuscated and Baritone is re-mapped as compile time. Processes in a scripting language might cause performance issues depending if they use reflection or not. What do you expect people to create with a scripting API?

**Comment by FriedrichDuerr** [comment]
What I meant with making it possible to inputs scripts it that also people that dont know much about any programming language have the chance to let baritone do more complicated things then walking somewhere. I do agree that processes are probably better suited but not everyone has the possibilities. Another argument is that writing a script would be faster then coding a process.

**Comment by IronException** [comment]
What I meant with making it possible to inputs scripts it that also people that dont know much about any programming language have the chance to let baritone do more complicated things then walking somewhere. I do agree that processes are probably better suited but not everyone has the possibilities. Another argument is that writing a script would be faster then coding a process.

**Comment by IronException** [comment]
the way I imagine it is that you can call the baritone processes in a row. So what baritone needs to do is look at what is the next thing to do and then lets that process run. If it is just that it would already help a lot of people.

Maybe if we want to go more complicated we could also let run different code when the process fails eg. Or other minecraft specific events... 
loops / goto might also come in handy.

**Comment by 0-x-2-2** [comment]
> What I meant with making it possible to inputs scripts it that also people that dont know much about any programming language have the chance to let baritone do more complicated things then walking somewhere. I do agree that processes are probably better suited but not everyone has the possibilities. Another argument is that writing a script would be faster then coding a process.

You would be coding a process but in a scripting language like js or lua,

**Comment by ZeroMemes** [comment]
![image](https://user-images.githubusercontent.com/24485393/64060096-ceb05280-cb8d-11e9-9529-bdbbef4c7525.png)
lol

**Comment by ZeroMemes** [comment]
NOT GONNA LIE
![image](https://user-images.githubusercontent.com/24485393/64060101-d8d25100-cb8d-11e9-9ed8-e3ae5e131185.png)

**Comment by 0-x-2-2** [comment]
> NOT GONNA LIE
> ![image](https://user-images.githubusercontent.com/24485393/64060101-d8d25100-cb8d-11e9-9ed8-e3ae5e131185.png)

Does not load with the Minecraft launcher but still something 😎

**Comment by ZeroMemes** [comment]
Probably would've if I wasn't doing poz meme to load it.

**Comment by LoganDark** [comment]
I think a DSL would be more better suited to this than trying to use Nashorn, or Lua or Python.

**Comment by SIGSTACKFAULT** [comment]
@ZeroMemes Don't tell me this is already a thing.

**Comment by ZeroMemes** [comment]
@SIGSTACKFAULT it’s not, I spent like 5 minutes setting it up on a local environment

**Comment by qiujiangkun** [comment]
I implemented Jython api for my own version of [Minecraft Robot](https://github.com/qiujiangkun/MineRobot) . It worked pretty well. Check it out. 
Nevertheless, implementing script or DSL may or may not be a painful process for such a huge project due to mapping and confusing.

**Comment by SIGSTACKFAULT** [comment]
> Pick a language (that doesn't suck)

Should we do a vote or something?

**Comment by IronException** [comment]
Im not sure why we need a language. I thought this is to make it possible for people that dont know much about programming to also do more complex things with baritone. So having to learn a script language doesnt help with the problem. In my opinion it should be possible to write the commands you would give baritone normaly in the script. Maybe if you have more experience a way could be added that fails are considered and other things like variables.

**Comment by LoganDark** [comment]
> should be possible to write the commands you would give baritone normaly line by line in the script

That would require another rewrite of the command system, and plus you would need some other operators as well. It could become like a bash shell, where in any context you can always chain commands and such. But I think the upcoming commands system is all we'll have for a while.

**Comment by IronException** [comment]
you just need an interpreter

**Comment by LoganDark** [comment]
To interpret the code and then do what? Commands only print messages to chat, they don't return even exit codes or anything. They just fire off and then do their own thing.

You can't just slap an interpreter on a completely unrelated system and just _tell_ it to work.

**Comment by IronException** [comment]
I think we think about the problem differently. Maybe this issue is also different from what I imagine.
The basic thing would be to pass each line in the script into ExampleBaritoneControl.runCommand (like chat commands are handled). When it fails it also prints that out in chat so you can detect fails in code ofc. The better question would be how to handle fails in the script. Oh and idk how you think about the script but I imagine it as a process of baritone that has a file as a parameter (or rly long string). (like how you give it a schematic to build). Then it can handle the commands without you needing to input them in game after each other

**Comment by LoganDark** [comment]
> The basic thing would be to pass each line in the script into ExampleBaritoneControl.runCommand (like chat commands are handled)

That's easy. The hard part is getting output from it, because commands don't work that way.

Even try this example:

```
#build <something>
#goal 0 0
#invert
#path
```

you would have to find a way for the build command to signal when it's complete, which would require a LOT of refactoring, not just in the commands system, but in mineprocess and etc.

**Comment by IronException** [comment]
ok I get it. Good point.

**Comment by IronException** [comment]
That interpreter also requires to communicate with the process to know when its done or failed

**Comment by LoganDark** [comment]
@IronException The command system needs to be build with that kind of support in mind, and the individual commands need to be redesigned to notify of success or failure. The interpreter can't just communicate with processes either because there's no way to know which one to communicate with.

**Comment by IronException** [comment]
okay I thought that wasnt that important but I definitly agree the command system should be changed
 I started doing something but then I heared that somebody is already doing it. So I guess we have to wait for that?

**Comment by LoganDark** [comment]
Yes, there will need to be major refactoring if you want an interpreter. I will say, however, that the new commands system will be slightly more flexible in this area

**Comment by SIGSTACKFAULT** [comment]
I didn't mean "add the ability to write a list of commands in a file and execute them one at a time".  

What I meant is, effectively, an API so Python-or-whatever can be used to tell baritone what to do. Loops, if statements, and all.

If you want to implement read-a-file-of-commands, you can do that to get the framework down for this.

**Comment by LoganDark** [comment]
> effectively, an API so Python-or-whatever can be used to tell baritone what to do. Loops, if statements, and all.

Oooorrrrr you can just use Java like everyone else? This has never been an issue.

**Comment by SIGSTACKFAULT** [comment]
Specifically an interpreted language than can be hot-reloaded at runtime.

**Comment by LoganDark** [comment]
I could probably do that, if you want

**Comment by 0-x-2-2** [comment]
[furry.txt](https://github.com/cabaletta/baritone/files/3693880/furry.txt)
🦍

**Comment by LoganDark** [comment]
@0-x-2-2 i call harassment

**Comment by SIGSTACKFAULT** [comment]
If you want to do file-full-of-commands-to-run-in-sequence, make another issue.

**Comment by LoganDark** [comment]
Oh no, that's not what I meant, but nevermind that, the offer's off the table. Baritone's contributors don't want me here

**Comment by 0-x-2-2** [comment]
>not a fan of harrassment
>>harasses people
pick one

**Comment by 0-x-2-2** [comment]
If you would have apologized to leijurv instead of being a child about it and just leaving everything would have been fine.

**Comment by LoganDark** [comment]
I tried apologizing, you all made a big deal about it and wouldn't stop telling me about how stupid it was so I left the group.

I post something completely unrelated in here and you come in with your "furry.txt" posting the chat logs for everyone to see.

I don't think I'm the child here.

**Comment by 0-x-2-2** [comment]
>I apologized multiple times, but you all made a big deal about it and wouldn't stop telling me about how stupid it was, so I left the group.

The chat logs say otherwise about you apologizing instead you just tried to act like you did nothing wrong and that you "warned him" even though from what you said leijurv just thought you were trying to send him something that cannot be sent over discord. 

>"furry.txt"

The `furry.txt` file was not named by me lol. 

>You've been hostile towards me from the very start anyway, so I think it's better for me to leave.

I pointed out flaws in your code and reasoning and maybe I was too harsh sorry for that.

**Comment by LoganDark** [comment]
@0-x-2-2 

```
WHAT?!
Last Tuesday at 5:17 PM
I'm sorry geez
```

```
WHAT?!
Last Tuesday at 5:05 PM
Holy shit
I didn't know
Sorry...
```

Nah fam have fun

**Comment by leijurv** [comment]
https://cdn.discordapp.com/attachments/541327608286019615/630228248634064936/Screenshot_20191005-192157.png
wait why did you block me on discord xD

what did i ever do to you? other than being slightly annoyed with you for some email spam (which btw i dont care much about at all)

**Comment by 0-x-2-2** [comment]
WHAT?!
Last Tuesday at 5:19 PM
Okay
I did warn you though

🦍

**Comment by leijurv** [comment]
`Baritone's contributors don't want me here, for being a "furry".` i'm not sure why you're transferring a problem with 0x to a problem with me? not sure if you noticed but it's brady and i that can push to baritone.... not 0x 

(0x resigned lmao owned)

**Comment by LoganDark** [comment]
@leijurv honestly I just blocked everyone on the baritone project, it's maybe not you personally, but you and brady both didn't seem very forgiving in the group chat

Dude, this isn't what issues are for

**Comment by shimakaze-chan-uwu** [comment]
WAAAAAAAAAH YOU DIDNT IMMEDIATELY FORGIVE ME
WAAAAAAAAAAAAAAAAAAAAAAH
👶 
🍼 
😿

**Comment by shimakaze-chan-uwu** [comment]
@LoganDark

**Comment by LoganDark** [comment]
Well at least I didn't make a big fuckin deal about it and start harassing you elsewhere, you're the reason I made a contribution to Baritone anyway so you're not that bad, just misguided

**Comment by leijurv** [comment]
all i've seen is 0x literally calling you a furry. you are a furry aren't you? i don't see the problem with him, for example, naming a file `furry.txt`?????????

also why should i care what 0x says to you or what you say to 0x just block him if you don't like it??

**Comment by shimakaze-chan-uwu** [comment]
Is This Guys Not A Furry? Or Does He Want Us To Call Him A Dog Or Cat Or Wolf Or Lion????

**Comment by acesblue** [comment]
> 
> 
> > The basic thing would be to pass each line in the script into ExampleBaritoneControl.runCommand (like chat commands are handled)
> 
> That's easy. The hard part is getting output from it, because commands don't work that way right now.
> 
> Even try this example:
> 
> ```
> #build <something>
> #goal 0 0
> #invert
> #path
> ```
> 
> you would have to find a way for the build command to signal when it's complete, which would require a LOT of refactoring, not just in the commands system, but in mineprocess and etc.

Actually the baritone says "building finished" and I can use AdvancedMacros to catch that keyphrase and continue the script.

**Comment by dominikaaaa** [comment]
Yes but the entire purpose of this issue is for the user to have access to the Baritone API in game, also someone else could say building finished in the chat and fuck you up.

**Comment by SylveonBottle** [comment]
Duplicate of #269, [according to l1ving](https://github.com/cabaletta/baritone/issues/269#issuecomment-751557525)

**Comment by ZacSharp** [comment]
I think something more high-level than the Java API would be a better fit. Having things in a scripting language could be used to hide the `onTick` calls and making the thing execute continuously with the time between ticks being spent in methods that make Baritone do things. So walking to a tree and mining all reachable logs could look roughly like this.
```java
pos = findNextTree()
walkTo(pos) // this can take quite some time
if (playerPosition() != pos) {
    log("Could not reach a tree!")
    exit()
}
logpos = findNearby("log")
while (logpos != null) {
    lookAt(logPos) // this one either takes one tick or is instant, not sure what would be better
    while (lookingAt() == logpos) {
        leftclick() // this takes just a single tick, also note that there is no need to release any keys manually
    }
    logpos = findNearby("log)
}
log("Done")
```
That way there is a real use to it for users which are probably more comfortable with thinking about behaviour as a continuous process and not as a dozen function calls per second. We could still expose the entire java api to it for advanced users who don't have a problem stumbling on obfuscated names.
For the implementation I'd suggest using the built in Java script engine api so users can just plug in their language of choice (jython, jlua, javascript, etc.) and on top of that we could use that api to provide ourselves with a script engine for a DSL, but I think a full language implementation is out of scope for Baritone, even if it's super simple. (This entire thing is possibly out of scope, in any variation)
